File: c/resistor-color-trio/resistor_color_trio.c
#include "resistor_color_trio.h"

resistor_value_t color_code(const resistor_band_t bands[]) {

    resistor_value_t resistor_value;
    resistor_value.unit = OHMS;

    // get the value of the first band
    resistor_value.value = bands[0] * 10;

    // get the value of the second band
    resistor_value.value += bands[1];

    // get the value of the third band
    resistor_value.value *= (unsigned int)pow(10, bands[2]);

    if (resistor_value.value >= 1000) {
        resistor_value.value /= 1000;
        resistor_value.unit = KILOOHMS;
    }

    return resistor_value;
}

File: ruby/resistor-color-trio/resistor_color_trio.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/resistor-color-trio
# Resistor Color Trio exercise
class ResistorColorTrio
  COLORS = %w[
    black
    brown
    red
    orange
    yellow
    green
    blue
    violet
    grey
    white
  ].freeze

  def initialize(bands)
    @bands = bands
  end

  def label
    value = resistor_value(@bands)
    value *= 10**COLORS.index(@bands[2])
    units = 'ohms'

    while value >= 1000
      value /= 1000
      units = 'kiloohms'
    end

    "Resistor value: #{value} #{units}"
  end

  def resistor_value(bands)
    resistor_value = COLORS.index(bands[0]) * 10

    resistor_value += COLORS.index(bands[1])

    resistor_value
  end

  private :resistor_value
end
